samples/bluetooth/hci_rpmsg/src/main.c

```diff
@@ -0,0 +1,455 @@
+		return -EINVAL;
+	}
+	net_buf_push_u8(buf, pkt_indicator);
+
+	LOG_HEXDUMP_DBG(buf->data, buf->len, "Final HCI buffer:");
+	rpmsg_send(&ep, buf->data, buf->len);
+
+	net_buf_unref(buf);
+
+	return 0;
+}
+
+#if defined(CONFIG_BT_CTLR_ASSERT_HANDLER)
+void bt_ctlr_assert_handle(char *file, u32_t line)
+{
+	LOG_ERR("Controller assert in: %s at %d", file, line);
+}
+#endif /* CONFIG_BT_CTLR_ASSERT_HANDLER */
+
+int endpoint_cb(struct rpmsg_endpoint *ept, void *data, size_t len, u32_t src,
+		void *priv)
+{
+	LOG_INF("Received message of %u bytes.", len);
+	hci_rpmsg_rx((u8_t *) data, len);
+
+	return RPMSG_SUCCESS;
+}
+
+static int hci_rpmsg_init(void)
+{
+	int err;
+	struct metal_init_params metal_params = METAL_INIT_DEFAULTS;
+
+	static struct virtio_vring_info   rvrings[2];
+	static struct virtio_device       vdev;
+	static struct rpmsg_device        *rdev;
+	static struct rpmsg_virtio_device rvdev;
+	static struct metal_io_region     *io;
+	static struct metal_device        *device;
+
+	/* Libmetal setup */
+	err = metal_init(&metal_params);
+	if (err) {
+		LOG_ERR("metal_init: failed - error code %d", err);
+		return err;
+	}
+
+	err = metal_register_generic_device(&shm_device);
+	if (err) {
+		LOG_ERR("Couldn't register shared memory device: %d", err);
+		return err;
+	}
+
+	err = metal_device_open("generic", SHM_DEVICE_NAME, &device);
+	if (err) {
+		LOG_ERR("metal_device_open failed: %d", err);
+		return err;
+	}
+
+	io = metal_device_io_region(device, 0);
+	if (!io) {
+		LOG_ERR("metal_device_io_region failed to get region");
+		return -ENODEV;
+	}
+
+	/* IPM setup */
+	ipm_tx_handle = device_get_binding("IPM_1");
+	if (!ipm_tx_handle) {
+		LOG_ERR("Could not get TX IPM device handle");
+		return -ENODEV;
+	}
+
+	ipm_rx_handle = device_get_binding("IPM_0");
+	if (!ipm_rx_handle) {
+		LOG_ERR("Could not get RX IPM device handle");
+		return -ENODEV;
+	}
+
+	ipm_register_callback(ipm_rx_handle, ipm_callback, NULL);
+
+	vq[0] = virtqueue_allocate(VRING_SIZE);
+	if (!vq[0]) {
+		LOG_ERR("virtqueue_allocate failed to alloc vq[0]");
+		return -ENOMEM;
+	}
+
+	vq[1] = virtqueue_allocate(VRING_SIZE);
+	if (!vq[1]) {
+		LOG_ERR("virtqueue_allocate failed to alloc vq[1]");
+		return -ENOMEM;
+	}
+
+	rvrings[0].io = io;
+	rvrings[0].info.vaddr = (void *)VRING_TX_ADDRESS;
+	rvrings[0].info.num_descs = VRING_SIZE;
+	rvrings[0].info.align = VRING_ALIGNMENT;
+	rvrings[0].vq = vq[0];
+
+	rvrings[1].io = io;
+	rvrings[1].info.vaddr = (void *)VRING_RX_ADDRESS;
+	rvrings[1].info.num_descs = VRING_SIZE;
+	rvrings[1].info.align = VRING_ALIGNMENT;
+	rvrings[1].vq = vq[1];
+
+	vdev.role = RPMSG_REMOTE;
+	vdev.vrings_num = VRING_COUNT;
+	vdev.func = &dispatch;
+	vdev.vrings_info = &rvrings[0];
+
+	/* setup rvdev */
+	err = rpmsg_init_vdev(&rvdev, &vdev, NULL, io, NULL);
+	if (err) {
+		LOG_ERR("rpmsg_init_vdev failed %d", err);
+		return err;
+	}
+
+	rdev = rpmsg_virtio_get_rpmsg_device(&rvdev);
+
+	err = rpmsg_create_ept(&ep, rdev, "k", RPMSG_ADDR_ANY,
```

Comment:
could you give an explicit name for the service ( e.g. "hci_bt")? it is an example, but it is a good practice to give a name in relation to the feature implemented, can help for debugging.

Reply:
yes, thank you, I changed it "bt_hci"